### PR TITLE
Add C++ core sysobject API

### PR DESCRIPTION
**Pull request by r-rojo**
* The cxx/core/sysobject wraps the Object API
* Add tests cases for cxx sysobject class
* Add some error checking in the Object API read/write routines
  * Check if the object can be read/written (if it's not a dir object)

### CODE REVIEW

File: testing/opae-cxx/test_object_cxx_core.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    ASSERT_GT(tokens_.size(), 0);
+    handle_ = handle::open(tokens_[0], 0);
+    ASSERT_NE(handle_.get(), nullptr);
+  }
+
+  virtual void TearDown() override {
+    system_->finalize();
+    tokens_.clear();
+    handle_->close();
+    handle_.reset();
```
Comment (ibarraza) [inline testing/opae-cxx/test_object_cxx_core.cpp]: Is this needed? Don't these get cleaned up by the destructor? And can we call reset on a closed handle?

File: common/include/opae/cxx/core/sysobject.h [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+  sysobject &operator=(const sysobject &o) = delete;
+
+  /**
+   * @brief Get a sysobject from a token. This will be read-only.
+   *
+   * @param t[in] Token object representing a resource.
+   * @param name[in] An identifier representing an object belonging to a
+   * resource
+   * represented by the token.
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: Move this line to end of previous.

File: common/include/opae/cxx/core/sysobject.h [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+   * @param[in] flags Control behavior of object identification and creation.
+   * FPGA_OBJECT_GLOB is used to indicate that the name should be treated as a
+   * globbing expression.  FPGA_OBJECT_RECURSE_ONE indicates that subobjects be
+   * created for objects one level down from the object identified by name.
+   * FPGA_OBJECT_RECURSE_ALL indicates that subobjects be created for all
+   * objects
+   * below the current object identified by name.
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: Move this line to end of previous.

File: common/include/opae/cxx/core/sysobject.h [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  static sysobject::ptr_t get(token::ptr_t t, const std::string &name,
+                              int flags = 0);
+
+  /**
+   * @brief Get a sysobject from a handle. This will be read-write.
+   *
+   * @param h[in] Handle object representing an open resource.
+   * @param name[in] An identifier representing an object belonging to a
+   * resource
+   * represented by the handle.
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: Move this line to end of previous.

File: common/include/opae/cxx/core/sysobject.h [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+   * @param[in] flags Control behavior of object identification and creation.
+   * FPGA_OBJECT_GLOB is used to indicate that the name should be treated as a
+   * globbing expression.  FPGA_OBJECT_RECURSE_ONE indicates that subobjects be
+   * created for objects one level down from the object identified by name.
+   * FPGA_OBJECT_RECURSE_ALL indicates that subobjects be created for all
+   * objects
+   * below the current object identified by name.
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: ditto

File: common/include/opae/cxx/core/sysobject.h [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+   *
+   * @return A shared_ptr to a sysobject instance.
+   */
+  static sysobject::ptr_t get(handle::ptr_t h, const std::string &name,
+                              int flags = 0);
+
+  /**
+   * @brief Get a sysobject from a object. This will be read-write if its
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: s/a object/an object/

File: common/include/opae/cxx/core/sysobject.h [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+   * parent has created from a handle..
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: s/has/was/

File: common/include/opae/cxx/core/sysobject.h [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+   *
+   * @param name[in] An identifier representing an object belonging this
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: s/belonging this/belonging to this/

File: common/include/opae/cxx/core/sysobject.h [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+   * object.
+   * @param[in] flags Control behavior of object identification and creation.
+   * FPGA_OBJECT_GLOB is used to indicate that the name should be treated as a
+   * globbing expression.  FPGA_OBJECT_RECURSE_ONE indicates that subobjects be
+   * created for objects one level down from the object identified by name.
+   * FPGA_OBJECT_RECURSE_ALL indicates that subobjects be created for all
+   * objects. Flags are defaulted to 0 meaning no flags.
+   * below the current object identified by name.
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: Copy/paste remnant?

File: common/include/opae/cxx/core/sysobject.h [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+   * @return A 64-bit value from the object.
+   */
+  uint64_t read64(int flags = 0) const;
+
+  /**
+   * @brief Write 64-bit value to an FPGA object.
+   * The value will be converted to string before writing. See flags below for
+   * changing that behavior.
+   *
+   * @param value The value to write to the object.
```
Comment (tswhison) [inline common/include/opae/cxx/core/sysobject.h]: s/@param/@param[in]/

File: testing/opae-cxx/test_object_cxx_core.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  std::vector<token::ptr_t> tokens_;
+  handle::ptr_t handle_;
+  test_platform platform_;
+  test_device invalid_device_;
+  test_system *system_;
+};
+
+TEST_P(sysobject_cxx_p, token_object) {
```
Comment (tswhison) [inline testing/opae-cxx/test_object_cxx_core.cpp]: Comments for the test cases?

**Comment by r-rojo** [comment]
Looks like this test is haunting us again:
image

**Pull request merged**